test/test_torch.py

```diff
@@ -4849,6 +4849,54 @@ def test_diff(self, device, dtype):
             scalar = torch.tensor(2, device=device, dtype=dtype)
             torch.diff(scalar)
 
+    def _test_histogram_numpy(self, t, bins, weights, density):
+        # Helper for test_histogram to compare with NumPy reference implementation
+        def to_np(t):
+            if t is None:
+                return None
+            if t.dtype == torch.bfloat16:
+                return t.to(dtype=torch.float, device="cpu").numpy()
+            else:
+                return t.cpu().numpy()
+
+        np_t = to_np(t)
+        np_bins = bins if isinstance(bins, int) else to_np(bins)
+        np_weights = to_np(weights)
+
+        (actual_hist, actual_bin_edges) = torch.histogram(t, bins, weight=weights, density=density)
+
+        expected = np.histogram(np_t, np_bins, weights=np_weights, density=density)
+        expected_hist = torch.from_numpy(expected[0])
+        expected_bin_edges = torch.from_numpy(expected[1])
+
+        self.assertEqual(actual_hist, expected_hist.to(actual_hist.dtype))
+        self.assertEqual(actual_bin_edges, expected_bin_edges.to(actual_bin_edges.dtype))
+
+    @onlyCPU
+    @dtypes(*(torch.testing.get_all_int_dtypes() + [torch.float32, torch.float64]))
+    def test_histogram(self, device, dtype):
+        shapes = (
+            (1,),
+            (1, 5),
+            (3, 5),
+            (1, 5, 1),
+            (2, 3, 5))
+
+        for contig in [True, False]:
```

Comment:
style: use one for loop over the itertools product of these values to avoid so much indentation